proposal/js-logger.d.ts

```diff
@@ -0,0 +1,206 @@
+ *
+ * ### Design Decisions
+ *
+ * #### Separate Logger Object
+ *
+ * This proposal defines a `Logger` interface that has its own `warn` and
+ * `debug` fields. Another alternative would have been to add both `warn` and
+ * `debug` to the `Options` interface directly. We chose to define a separate
+ * interface to open the possibility for libraries to expose re-usable `Logger`
+ * objects, or even to wrap `Logger`s with decorator functions.
+ */
+
+/* ## API */
+
+import {URL} from 'url';
+
+import '../spec/js-api';
+
+declare module '../spec/js-api' {
+  interface _Options {
+    /**
+     * An object that provides callbacks for the compiler to use in lieu of its
+     * default messaging behavior.
+     *
+     * The compiler must treat an `undefined` logger identically to an object
+     * that doesn't have `warn` or `debug` fields.
+     */
+    logger?: Logger;
+  }
+}
+
+/**
+ * An object that provides callbacks for handling messages from the
+ * compiler.
+ */
+export interface Logger {
+  /**
+   * If this field is defined, the compiler must invoke it under the following
+   * circumstances:
+   *
+   * * When it encounters a `@warn` rule:
+   *   * Let `value` be the result of evaluating the rule's expression.
+   *   * Let `message` be `value`'s text if it's a string, or the result of
+   *     serializing `value` if it's not.
+   *   * Invoke `warn` with `message` and an object with `deprecation` set to
+   *     `false` and `stack` set to a string representation of the current Sass
+   *     stack trace.
+   *
+   *     > The specific format of the stack trace may vary from implementation
+   *     > to implementation.
+   *
+   * * When it encounters anything else that the user needs to be warned about:
+   *
+   *   > This is intentionally vague about what counts as a warning.
+   *   > Implementations have a considerable degree of flexibility in defining
+   *   > this for themselves, although in some cases warnings are mandated by
+   *   > the specification (such as in preparation for a breaking change).
+   *
+   *   * Let `options` be an empty object.
+   *   * If this warning is caused by behavior that used to be allowed but will
+   *     be disallowed in the future, set `options.deprecation` to `true`.
+   *     Otherwise, set `options.deprecation` to `false`.
+   *   * If this warning is associated with a specific span of a Sass
+   *     stylesheet, set `options.span` to a `SourceSpan` that covers that span.
+   *   * If this warning occurred during execution of a stylesheet, set
+   *     `options.stack` to a string representation of the current Sass stack
+   *     trace.
+   *   * Invoke `warn` with a string describing the warning and `options`.
+   *
+   * If this field is defined, the compiler must not surface warnings in any way
+   * other than inokving `warn`.
+   */
+  warn?(
+    message: string,
+    options: {
+      deprecation: boolean;
```

Comment:
Could this be optional and default to `false`?

Reply:
Since this is an argument that's being passed *to* user code, I'd rather have it always set so that callbacks don't need to think about the distinction between `undefined` and `false`.